File: src/main/scala/homework3/java/PurchaseAnalytic.java
package homework3.java;

import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.stream.Collectors;

public class PurchaseAnalytic {
    @Data
    @AllArgsConstructor
    static class Product {
        String name;
        int price;
    }

    @Data
    @AllArgsConstructor
    static class Purchase {
        String customer;
        Product product;
    }

    public static void main(String[] args) {
        String[] customersList = {"Robert", "Anna", "Sarah", "Moshe", "Benjamin", "Anat", "Dan", "Mark", "Sofia"};

        Product[] productList = {
                new Product("Salt", 10),
                new Product("Sugar", 30),
                new Product("Oil", 35),
                new Product("Bread", 15),
                new Product("Flour", 22),
                new Product("Milk", 11),
                new Product("Tea", 28),
                new Product("Coffee", 39),
                new Product("Cheese", 42),
                new Product("Sausage", 51)
        };

        Random rnd = new Random();
        int purchasesNumber = 200;

// GENERATE RANDOM PURCHASES LIST
        List<Purchase> purchases = rnd.ints(purchasesNumber).
                mapToObj(n -> new Purchase(customersList[rnd.nextInt(customersList.length)], productList[rnd.nextInt(productList.length)])).
                collect(Collectors.toList());

        getCustomersWithList(purchases).
                entrySet().
                forEach(System.out::println);

        getCustomersWithSum(purchases).
                entrySet().
                forEach(System.out::println);

        getCustomersWithAverage(purchases).
                entrySet().
                forEach(System.out::println);
    }

    static Map<String, Double> getCustomersWithAverage(List<Purchase> purchases) {
        return purchases.stream().
                collect(Collectors.groupingBy(p -> p.getCustomer(), Collectors.averagingInt(p -> p.getProduct().getPrice())));
    }

    static Map<String, Integer> getCustomersWithSum(List<Purchase> purchases) {
        return purchases.stream().
                collect(Collectors.groupingBy(p -> p.getCustomer(), Collectors.summingInt(p -> p.getProduct().getPrice())));
    }

    static Map<String, List<Purchase>> getCustomersWithList(List<Purchase> purchases) {
        return purchases.stream().
                collect(Collectors.groupingBy(p -> p.getCustomer()));
    }
}

File: src/main/scala/homework3/scala/PurchaseAnalytic.scala
package homework3.scala

import scala.util.Random

object PurchaseAnalytic {

  case class Product(name: String, price: Int){}

  case class Purchase(customer: String, product: Product){}

  val customersList: Array[String] = Array("Robert", "Anna", "Sarah", "Moshe", "Benjamin", "Anat", "Dan", "Mark", "Sofia")

  val productList: Array[Product] = Array(
    Product("Salt", 10),
    Product("Sugar", 30),
    Product("Oil", 35),
    Product("Bread", 15),
    Product("Flour", 22),
    Product("Milk", 11),
    Product("Tea", 28),
    Product("Coffee", 39),
    Product("Cheese", 42),
    Product("Sausage", 51))

  val purchasesNumber = 50

// GENERATE RANDOM PURCHASES LIST
  val purchases = LazyList.range(0, purchasesNumber, 1).
    map(n => Purchase(customersList(Random.nextInt(customersList.length)), productList(Random.nextInt(productList.length)))).
    toList

  def main(args: Array[String]): Unit = {

    getCustomersWithList.foreach(println)

    getCustomersWithSum.foreach(println)

    getCustomersWithAverage.foreach(println)

  }

  def getCustomersWithList: Map[String, List[Product]] = purchases.
    groupMap(p => p.customer)(p => p.product)

  def getCustomersWithSum: Map[String, Int] = purchases.
    groupMap(p => p.customer)(p => p.product.price).map(tup => (tup._1, tup._2.sum))

  def getCustomersWithAverage: Map[String, Float] = purchases.
    groupMap(p => p.customer)(p => p.product.price).map(tup => (tup._1, tup._2.sum.asInstanceOf[Float]/tup._2.length))

}
